### ISSUE TITLE
Seed all sources of randomness for reproducible pipeline runs

### ISSUE THREAD

**Issue by zachtheyek** [zachtheyek/Aetherscan#49]
## Problem

Our pipeline introduces randomness at multiple stages — data generation, training data shuffling, VAE sampling, weight initialization — but most of these are **not seeded**. This means that two runs with identical configuration and data will produce different results, making it impossible to:

- Debug training regressions (can't distinguish code changes from random variation)
- Compare experiments fairly (apples-to-oranges when randomness differs)
- Reproduce a specific result for validation or publication

The goal is **not** to eliminate randomness — randomness is essential for generalization (shuffling), realistic data augmentation (signal injection), and the VAE's latent space (reparameterization). The goal is to make randomness **deterministic given a seed**, so that any run can be exactly reproduced by reusing the same seed.

---

### Already Seeded (Reproducible)

**sklearn Random Forest** — `src/aetherscan/models/random_forest.py`
- `RandomForestClassifier(random_state=config.rf.seed)` (line 64) — model training
- `shuffle(features, labels, random_state=config.rf.seed)` (line 89) — pre-training data shuffle

Both use `config.rf.seed` (default `11`, overridable via `--rf-seed`). This is handled correctly.

#### 1. Data Generation — `src/aetherscan/data_generation.py` (High Impact)

Worker processes seed with `os.getpid()` (lines 89-90), which differs across runs. All downstream `random.random()` and `np.random.choice()` calls in `new_cadence()`, `create_false()`, `create_true_single()`, and `create_true_double()` are affected — this controls starting frequency bins, drift directions, drift rate noise, signal widths, background plate selection, and SNR values. **This is the largest source of non-reproducibility since it shapes the training data itself.**

Affected lines: 181, 187, 194, 202, 229, 355, 380, 436, 455, 507, 526

#### 2. Training Data Shuffling — `src/aetherscan/train.py` (Medium Impact)

`np.random.shuffle(indices)` (line 433) in `prepare_distributed_datasets()` shuffles training indices each epoch using the global numpy RNG. Affects the order samples are fed to the model.

#### 3. VAE Sampling Layer — `src/aetherscan/models/vae.py` (Medium Impact)

`tf.keras.backend.random_normal()` (line 42) in the `Sampling` layer uses TensorFlow's default RNG. No `tf.random.set_seed()` is called anywhere in the codebase.

#### 4. Keras Weight Initialization — `src/aetherscan/models/vae.py` (Medium Impact)

`HeNormal()` (~10 locations) and `GlorotNormal()` (~3 locations) draw random initial weights. Without a TF-level seed, initialization differs across runs.

#### 5. Plot Subsampling — `src/aetherscan/train.py` (Low Impact)

`np.random.choice()` (line 3040) in `plot_injection_stats_biases()`. Only affects visualization, not training — but still non-reproducible.

---

### Add a global `seed` to config

Add a top-level `seed: int = 42` to the training config (separate from the existing `rf.seed`). This serves as the base seed for the entire pipeline. Individual components derive their seeds from it.

### Per-component seeding strategy

| Area | Current State | Fix |
|------|--------------|-----|
| **Data generation workers** | `random.seed(os.getpid())` / `np.random.seed(os.getpid())` | Seed with `base_seed + worker_id` instead of PID. Pass the base seed into the worker init function and combine with a deterministic worker index. |
| **Data generation calls** | Unseeded `random.random()`, `np.random.choice()` | After fixing worker seeding above, these become deterministic within each worker (they inherit the worker's seeded state). No changes needed to individual call sites. |
| **Training batch shuffle** | Unseeded `np.random.shuffle()` | Use a local `np.random.default_rng(seed)` instance instead of the global numpy RNG. |
| **TensorFlow (VAE + weight init)** | No TF seed set | Call `tf.random.set_seed(seed)` early in the training pipeline, before model construction. Covers both weight initialization and the sampling layer. |
| **Plot subsampling** | Unseeded `np.random.choice()` | Use a local `np.random.default_rng(seed)` instance. |

### Key principles

- **Use `np.random.default_rng()` over `np.random.seed()`** — NumPy recommends the newer Generator API. It avoids polluting global state and makes seed ownership explicit.
- **Derive component seeds from the base seed** — e.g., `tf.random.set_seed(base_seed)`, `rng = np.random.default_rng(base_seed + 1)`, workers get `base_seed + worker_id`. This prevents identical streams across components while keeping everything deterministic.
- **Don't remove randomness** — every `random.random()`, `shuffle()`, and `random_normal()` call stays. We're only making the random sequences reproducible given the same seed.

### Optional: Full TF GPU determinism

For exact GPU reproducibility, TensorFlow also needs:
```python
tf.config.experimental.enable_op_determinism()
```
This forces deterministic GPU ops at the cost of some performance. Worth considering but may not be necessary if approximate reproducibility (from seeding alone) is sufficient.

---

## Summary

| Priority | Component | Impact |
|----------|-----------|--------|
| High | Data generation worker seeding | Shapes training data |
| Medium | Training batch shuffle | Affects sample ordering |
| Medium | TF global seed (VAE + weight init) | Affects model initialization and latent sampling |
| Low | Plot subsampling | Visualization only |

**Comment by zachtheyek** [comment]
note, the above issue was generated by claude. its claims have not been fully verified. pls do so before breaking ground
